### PR TITLE
Added PAN token to defaults.json, added coingeckoid property, impleme…

### PR DESCRIPTION
**Pull request by arbreton**
…nted coingecko API, for PAN usage, prepared for future usage

### CODE REVIEW

**Comment by aminlatifi** [comment]
@arbreton 
* The code has eslint problems, fix those please.
* As I understood you fetch the price from today 00:00:00  to the timestamp the rate is requested. What if timestamp is sometime in the middle of today?
* Code walks through the rates to find the closest time to timestamp. We cannot fetch the hourly rates and find the exact match? Can we archive the rates (not just for the timestamp) to use sometimes latter?
* The coinGecko logic is implemented in `_getRatesCryptocompare` function, it's not cool and it should be implemented in separate function with related name.

**Comment by arbreton** [comment]
> @arbreton
> 
> * The code has eslint problems, fix those please.
> * As I understood you fetch the price from today 00:00:00  to the timestamp the rate is requested. What if timestamp is sometime in the middle of today?
> * Code walks through the rates to find the closest time to timestamp. We cannot fetch the hourly rates and find the exact match? Can we archive the rates (not just for the timestamp) to use sometimes latter?
> * The coinGecko logic is implemented in `_getRatesCryptocompare` function, it's not cool and it should be implemented in separate function with related name.

Updated the code base, moved coinGecko implementation to a proper function, fixed ESLint problems

Answering the following questions right now seems that we don't have a way to request various timestamps because we only have a date picker on the milestone creation window
Our current base is not enabled for daily rates to my understanding at this point, it will be an interesting feature

File: src/services/conversionRates/getConversionRatesService.js [hunk 1]
```diff
@@ -24,6 +24,56 @@ const _getRatesDb = async (app, timestamp, symbol = 'ETH') => {
...
 
+/**
+ * Get the rates from the coingecko API
+ *
+ * @throws Error if fetching the rates from cryptocompare API failed
+ *
+ * @param {Number} timestamp   Timestamp for which the value should be retrieved
+ * @param {Array}  ratestToGet Rates that are missing in the DB and should be retrieved
+ * @param {Array}  stableCoins coins whose value equal one usd
+ *
```
Comment (aminlatifi) [inline src/services/conversionRates/getConversionRatesService.js:35]: These lines are copied from `_getRatesCryptocompare` function doc and are not correct

File: src/services/conversionRates/getConversionRatesService.js [hunk 2]
```diff
@@ -47,13 +97,17 @@ const _getRatesCryptocompare = async (timestamp, ratesToGet, symbol, stableCoins
...
     if (rateSymbol !== requestSymbol) {
-      const resp = JSON.parse(
-        await rp(
-          `https://min-api.cryptocompare.com/data/dayAvg?fsym=${requestSymbol}&tsym=${rateSymbol}&toTs=${timestampMS}&extraParams=giveth`,
-        ),
-      );
-
-      if (resp && resp[rateSymbol]) rates[r] = resp[rateSymbol];
+      if (requestSymbol === 'PAN') {
+        rates[r] = await _getRatesCoinGecko(rateSymbol, timestampMS, coingeckoId);
```
Comment (aminlatifi) [inline src/services/conversionRates/getConversionRatesService.js:101]: It's still called inside `_getRatesCryptocompare` and not cool

File: src/services/conversionRates/getConversionRatesService.js [hunk 3]
```diff
@@ -24,6 +24,56 @@ const _getRatesDb = async (app, timestamp, symbol = 'ETH') => {
...
+    let difference = 0;
+    let bestIndex = 0;
+    let bestDifference = Infinity;
+    let i;
+    let cur;
+    let priceTime;
+
+    const { prices } = testRep;
+
+    for (i = 0; i < prices.length; i += 1) {
```
Comment (aminlatifi) [inline src/services/conversionRates/getConversionRatesService.js:60]: use forEach

File: src/services/conversionRates/getConversionRatesService.js [hunk 4]
```diff
@@ -24,6 +24,56 @@ const _getRatesDb = async (app, timestamp, symbol = 'ETH') => {
...
+      cur = prices[i];
```
Comment (aminlatifi) [inline src/services/conversionRates/getConversionRatesService.js:61]: Use array destructuring please

File: src/services/conversionRates/getConversionRatesService.js [hunk 5]
```diff
@@ -24,6 +24,56 @@ const _getRatesDb = async (app, timestamp, symbol = 'ETH') => {
...
+      const curRounded = Math.round(cur[0] / 1000);
```
Comment (aminlatifi) [inline src/services/conversionRates/getConversionRatesService.js:62]: curRound can be removed

File: src/services/conversionRates/getConversionRatesService.js [hunk 6]
```diff
@@ -24,6 +24,56 @@ const _getRatesDb = async (app, timestamp, symbol = 'ETH') => {
...
+ */
+const _getRatesCoinGecko = async (rateSymbol, timestampMS, coingeckoId) => {
+  const today = new Date(new Date().setUTCDate(new Date().getUTCDate()));
+  const todayUTC = today.setUTCHours(0, 0, 0, 0);
+  const todayRounded = Math.round(todayUTC / 1000);
+  const differenceDays = (todayRounded - timestampMS) / 86400;
+  const daysMax = Math.round(differenceDays);
+  const testRep = JSON.parse(
+    await rp(
+      `https://api.coingecko.com/api/v3/coins/${coingeckoId}/market_chart?vs_currency=${rateSymbol}&days=${daysMax}`,
```
Comment (aminlatifi) [inline src/services/conversionRates/getConversionRatesService.js:46]: Let's use `/coins/{id}/market_chart/range` api for efficiency. Send request for some finite range before and after the requested timestamp, 10 minutes maybe enough.

File: src/services/conversionRates/getConversionRatesService.js [hunk 7]
```diff
@@ -148,6 +202,18 @@ const getConversionRates = async (app, requestedDate, requestedSymbol = 'ETH') =
 
   const fiat = app.get('fiatWhitelist');
   const stableCoins = app.get('stableCoins') || [];
+  const tokens = app.get('activeTokenWhitelist');
+  let coingeckoId = '';
+
+  const keys = Object.keys(tokens);
```
Comment (aminlatifi) [inline src/services/conversionRates/getConversionRatesService.js:208]: tokens is an array... it's inappropriate? Object.keys or Object.values are used for iterating over objects. Use other standard iterating over array.

**Review by aminlatifi — CHANGES_REQUESTED** [review]

File: src/services/conversionRates/getConversionRatesService.js [hunk 8]
```diff
@@ -24,6 +24,56 @@ const _getRatesDb = async (app, timestamp, symbol = 'ETH') => {
...
+ * @throws Error if fetching the rates from cryptocompare API failed
+ *
+ * @param {Number} timestamp   Timestamp for which the value should be retrieved
+ * @param {Array}  ratestToGet Rates that are missing in the DB and should be retrieved
+ * @param {Array}  stableCoins coins whose value equal one usd
+ *
+ * @return {Object} Rates object in format { EUR: 241, USD: 123 }
+ */
+const _getRatesCoinGecko = async (rateSymbol, timestampMS, coingeckoId) => {
+  const today = new Date(new Date().setUTCDate(new Date().getUTCDate()));
```
[the remark this answers is not in the document]
Comment (arbreton) [reply src/services/conversionRates/getConversionRatesService.js:39]: Really? I think that's the way it is on line 192, we should change it as well maybe

File: src/services/conversionRates/getConversionRatesService.js [hunk 9]
```diff
@@ -24,6 +24,56 @@ const _getRatesDb = async (app, timestamp, symbol = 'ETH') => {
...
+  const todayUTC = today.setUTCHours(0, 0, 0, 0);
+  const todayRounded = Math.round(todayUTC / 1000);
+  const differenceDays = (todayRounded - timestampMS) / 86400;
```
[the remark this answers is not in the document]
Comment (arbreton) [reply src/services/conversionRates/getConversionRatesService.js:42]: That's interesting and a good observation, i think i have an idea!

File: src/services/conversionRates/getConversionRatesService.js (hunk 1 above)
Comment (arbreton) [reply src/services/conversionRates/getConversionRatesService.js:35]: Fixed this block

File: src/services/conversionRates/getConversionRatesService.js (hunk 2 above)
Comment (arbreton) [reply src/services/conversionRates/getConversionRatesService.js:101]: fixed

File: src/services/conversionRates/getConversionRatesService.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (aminlatifi) [reply src/services/conversionRates/getConversionRatesService.js:42]: Line 192 is different, it has minus one

File: src/services/conversionRates/getConversionRatesService.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (aminlatifi) [reply src/services/conversionRates/getConversionRatesService.js:39]: Line 192 is different, it has minus one. Also setUTCDate returns number and not date

File: src/services/conversionRates/getConversionRatesService.js (hunk 5 above)
Comment (arbreton) [reply src/services/conversionRates/getConversionRatesService.js:62]: Fixed

File: src/services/conversionRates/getConversionRatesService.js (hunk 3 above)
Comment (arbreton) [reply src/services/conversionRates/getConversionRatesService.js:60]: fixed

File: src/services/conversionRates/getConversionRatesService.js (hunk 4 above)
Comment (arbreton) [reply src/services/conversionRates/getConversionRatesService.js:61]: How specifically? not sure

File: src/services/conversionRates/getConversionRatesService.js (hunk 6 above)
Comment (arbreton) [reply src/services/conversionRates/getConversionRatesService.js:46]: Didn't understood this

File: src/services/conversionRates/getConversionRatesService.js (hunk 7 above)
Comment (arbreton) [reply src/services/conversionRates/getConversionRatesService.js:208]: Not sure what you mean?, will need an example

**Comment by arbreton** [comment]
I think it got merged sooner than expected, if something need to be changed we can definitely update it, so we can release PAN !! wooho!

**Comment by aminlatifi** [comment]
Reverted the merge. @arbreton Please fix the problems mentioned in review. If you need any help let me know.

File: src/services/conversionRates/getConversionRatesService.js (hunk 6 above)
Comment (arbreton) [reply src/services/conversionRates/getConversionRatesService.js:46]: Oh i understand this one, i will check it out!

File: src/services/conversionRates/getConversionRatesService.js (hunk 4 above)
Comment (arbreton) [reply src/services/conversionRates/getConversionRatesService.js:61]: This i'm not sure yet

File: src/services/conversionRates/getConversionRatesService.js (hunk 7 above)
Comment (arbreton) [reply src/services/conversionRates/getConversionRatesService.js:208]: What would be the alternative?

**Pull request merged**